src/presto.py

```diff
@@ -452,27 +450,30 @@ def forward(
         all_tokens.append(tokens)
 
         # now we calculate the mask for these [b, t] tokens
-        group_mask = repeat(
-            dynamic_world == DynamicWorld2020_2021.class_amount, "b t -> b t d", d=tokens.shape[-1]
-        )
+        group_mask = dynamic_world == DynamicWorld2020_2021.class_amount
         all_masks.append(group_mask)
 
         x = torch.cat(all_tokens, dim=1)  # [batch, timesteps, embedding_dim]
-        mask = torch.cat(all_masks, dim=1)  # [batch, timesteps, embedding_dim]
-        x, kept_indices, removed_indices = self.mask_tokens(x, mask)
+        mask = torch.cat(all_masks, dim=1)  # [batch, timesteps]
+        x, orig_indices, upd_mask = self.mask_tokens(x, mask)
 
         # append latlon tokens
         latlon_tokens = self.latlon_embed(self.cartesian(latlons)).unsqueeze(1)
         x = torch.cat((latlon_tokens, x), dim=1)
+        upd_mask = torch.cat((torch.zeros(x.shape[0])[:, None].to(device), upd_mask), dim=1)
+        orig_indices = torch.cat(
+            (torch.zeros(x.shape[0])[:, None].to(device).int(), orig_indices + 1),
+            dim=1,
+        )
 
         # apply Transformer blocks
         for blk in self.blocks:
-            x = blk(x)
+            x = blk(x, attn_mask=~upd_mask.bool())
```

Comment:
@rubencart, what happens here if I have some values of `x` which are masked?

i.e. `x` has shape `[b, t, 128]` for `t` (mostly) unmasked tokens. What happens if for `b=2`, `t=1` is masked?

I am specifically worried that when we return `self.norm(x.mean(dim=1))` on line 475 below, we will be including tokens which would have otherwise been removed.

Reply:
The attention ignores them. But yes we have to exclude them from the mean, good catch